agentendpoint/patch_windows.go

```diff
@@ -52,53 +52,65 @@ func classFilter(cs []agentendpointpb.WindowsUpdateSettings_Classification) ([]s
 	return cf, nil
 }
 
-func (r *patchTask) wuaUpdates(ctx context.Context) error {
+func (r *patchTask) installWUAUpdates(ctx context.Context, cf []string) (int32, error) {
+	r.infof("Searching for available Windows updates.")
 	session, err := packages.NewUpdateSession()
 	if err != nil {
-		return err
+		return 0, err
 	}
 	defer session.Close()
 
+	updts, err := ospatch.GetWUAUpdates(session, cf, r.Task.GetPatchConfig().GetWindowsUpdate().GetExcludes(), r.Task.GetPatchConfig().GetWindowsUpdate().GetExclusivePatches())
+	if err != nil {
+		return 0, err
+	}
+
+	count, err := updts.Count()
+	if err != nil {
+		return 0, err
+	}
+
+	if count == 0 {
+		return 0, nil
+	}
+
+	r.infof("%d Windows updates to install", count)
+
+	for i := 0; i < int(count); i++ {
+		if err := r.reportContinuingState(ctx, agentendpointpb.ApplyPatchesTaskProgress_APPLYING_PATCHES); err != nil {
+			return 0, err
+		}
+		updt, err := updts.Item(i)
+		if err != nil {
+			return 0, err
+		}
+		defer updt.Release()
+
+		if err := session.InstallWUAUpdate(updt); err != nil {
+			return 0, fmt.Errorf(`installUpdate(class, excludes, updt): %v`, err)
+		}
+	}
+
+	return count, nil
+}
+
+func (r *patchTask) wuaUpdates(ctx context.Context) error {
 	cf, err := classFilter(r.Task.GetPatchConfig().GetWindowsUpdate().GetClassifications())
 	if err != nil {
 		return err
 	}
 
 	// We keep searching for and installing updates until the count == 0 or there is an error.
-	retries := 50
+	retries := 20
```

Comment:
what's the reason that we have to retry; is it that we do not get all updates in one run or is it like the windows update installation are flaky? asking because retries would increase the time to patch an instance

Reply:
Updates are flaky :)

Reply:
ok